File: goatx402-canton-demo/src/lib/x402.ts
// x402 envelope discovery + accepts selection (PLAN.md §5.3 + §6.9).
//
// The merchant returns `402 Payment Required` with a JSON body of shape:
//   {
//     "x402Version": 1,
//     "accepts": [{ scheme, amount, currency, trustedIssuer, payTo,
//                   facilitator, resource, merchantRequestId }],
//     "error": "payment_required"
//   }
// The SPA picks the first `canton-daml` entry; we keep selection logic
// deliberately small so the browser flow mirrors client-cli.

export interface AcceptEntry {
  scheme: string;
  amount: string;
  currency: string;
  trustedIssuer: string;
  payTo: string;
  facilitator: string;
  resource: string;
  merchantRequestId: string;
}

export interface X402Envelope {
  x402Version: number;
  accepts: AcceptEntry[];
  error?: string;
}

export class X402DiscoveryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "X402DiscoveryError";
  }
}

// fetch402 calls the merchant `/resource` without `X-PAYMENT`. Returns the
// parsed envelope if the status is 402, throws otherwise (the merchant must
// challenge before we can pay).
export async function fetch402(
  merchantURL: string,
  resourcePath: string,
  fetchImpl: typeof fetch = fetch,
): Promise<X402Envelope> {
  const url = merchantURL + resourcePath;
  const resp = await fetchImpl(url, { method: "GET" });
  if (resp.status !== 402) {
    throw new X402DiscoveryError(
      `expected 402 from ${url}, got ${resp.status}`,
    );
  }
  const ct = resp.headers.get("content-type") ?? "";
  if (!ct.toLowerCase().includes("application/json")) {
    throw new X402DiscoveryError(
      `402 response from ${url} is not JSON (content-type=${ct})`,
    );
  }
  const body = (await resp.json()) as X402Envelope;
  if (!body || !Array.isArray(body.accepts)) {
    throw new X402DiscoveryError("402 envelope missing accepts[]");
  }
  return body;
}

// selectCantonDaml returns the first accepts entry whose scheme is
// "canton-daml". Throws if none are present — the SPA only knows how to pay
// over Canton.
export function selectCantonDaml(env: X402Envelope): AcceptEntry {
  const entry = env.accepts.find((a) => a.scheme === "canton-daml");
  if (!entry) {
    throw new X402DiscoveryError(
      "402 envelope has no canton-daml accepts entry",
    );
  }
  if (!entry.merchantRequestId) {
    throw new X402DiscoveryError(
      "canton-daml accepts entry missing merchantRequestId",
    );
  }
  if (!entry.payTo || !entry.amount || !entry.currency || !entry.trustedIssuer) {
    throw new X402DiscoveryError(
      "canton-daml accepts entry missing required fields",
    );
  }
  return entry;
}

File: goatx402-canton-cli/internal/x402/x402.go
// Package x402 discovers a 402 Payment Required envelope from a merchant
// and selects the `canton-daml` accepts entry per PLAN.md §5.3.
package x402

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// SchemeCantonDaml is the only accepts.scheme the CLI handles.
const SchemeCantonDaml = "canton-daml"

// Envelope is the parsed 402 body. Unknown fields are ignored so the merchant
// can grow the envelope shape (e.g. add `extensions`) without breaking the CLI.
type Envelope struct {
	X402Version int      `json:"x402Version"`
	Accepts     []Accept `json:"accepts"`
	Error       string   `json:"error"`
}

// Accept mirrors §5.3 accepts[*]. The CLI uses the canton-daml entry; other
// schemes are passed through but never selected.
type Accept struct {
	Scheme            string `json:"scheme"`
	Amount            string `json:"amount"`
	Currency          string `json:"currency"`
	TrustedIssuer     string `json:"trustedIssuer"`
	PayTo             string `json:"payTo"`
	Facilitator       string `json:"facilitator"`
	Resource          string `json:"resource"`
	MerchantRequestID string `json:"merchantRequestId"`
}

// Discover issues a GET against merchantURL+resourcePath, expects a 402, and
// returns the parsed envelope. Any 2xx is a contract violation (the merchant
// should always 402 before payment) and surfaces as an error so the caller
// can decide what to do.
func Discover(ctx context.Context, hc *http.Client, merchantURL, resourcePath string) (Envelope, *http.Response, error) {
	u := strings.TrimRight(merchantURL, "/") + ensureSlash(resourcePath)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return Envelope{}, nil, fmt.Errorf("x402: build request: %w", err)
	}
	resp, err := hc.Do(req)
	if err != nil {
		return Envelope{}, nil, fmt.Errorf("x402: GET %s: %w", u, err)
	}
	if resp.StatusCode != http.StatusPaymentRequired {
		body, _ := io.ReadAll(resp.Body)
		_ = resp.Body.Close()
		return Envelope{}, resp, fmt.Errorf(
			"x402: expected 402 from %s, got %d: %s",
			u, resp.StatusCode, truncate(string(body), 256),
		)
	}
	defer resp.Body.Close()
	var env Envelope
	if err := json.NewDecoder(resp.Body).Decode(&env); err != nil {
		return Envelope{}, resp, fmt.Errorf("x402: decode 402 body: %w", err)
	}
	return env, resp, nil
}

// SelectCantonDaml returns the first canton-daml accepts entry. Returns an
// error if no canton-daml entry is present.
func SelectCantonDaml(env Envelope) (Accept, error) {
	for _, a := range env.Accepts {
		if a.Scheme == SchemeCantonDaml {
			return a, nil
		}
	}
	return Accept{}, fmt.Errorf("x402: no %q accepts entry in 402 envelope", SchemeCantonDaml)
}

func ensureSlash(p string) string {
	if p == "" {
		return "/"
	}
	if strings.HasPrefix(p, "/") {
		return p
	}
	return "/" + p
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "…"
}
